### DISCUSSION TITLE
error compiling test: no matching function for call to ‘isinf(const MetaPhysicL

### DISCUSSION THREAD

**Discussion by dealmeidavf** [idaholab/moose#20597] [Q&A General] [answered]
Hello,
In a recent MOOSE pull from repo, and full recompile (GCC upgraded to 11.2.1) on Linux, the test build had an error:

```c++
/usr/include/eigen3/Eigen/src/Core/MathFunctions.h:1004:36: error: no matching function for call to ‘isinf(const MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned int, MetaPhysicL::NWrapper<53> >, true>&)’
 1004 |     return isinf EIGEN_NOT_A_MACRO (x);
      |            ~~~~~~~~~~~~~~~~~~~~~~~~^~~
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/11.2.1/include/g++-v11/complex:44,
                 from /home/dealmeida/gentoo-home/work/codes/moose/scripts/../libmesh/installed/include/libmesh/libmesh_common.h:46,
                 from /home/dealmeida/gentoo-home/work/codes/moose/framework/build/header_symlinks/DualReal.h:14,
                 from /home/dealmeida/gentoo-home/work/codes/moose/framework/build/header_symlinks/DualRealOps.h:13,
                 from /home/dealmeida/gentoo-home/work/codes/moose/framework/src/utils/RankTwoTensor.C:10:
/usr/lib/gcc/x86_64-pc-linux-gnu/11.2.1/include/g++-v11/cmath:605:5: note: candidate: ‘template<class _Tp> constexpr typename __gnu_cxx::__enable_if<std::__is_integer<_Tp>::__value, bool>::__type std::isinf(_Tp)’
  605 |     isinf(_Tp __x)
      |     ^~~~~
/usr/lib/gcc/x86_64-pc-linux-gnu/11.2.1/include/g++-v11/cmath:605:5: note:   template argument deduction/substitution failed:

[snip]

make: *** [/home/dealmeida/gentoo-home/work/codes/moose/framework/build.mk:145: /home/dealmeida/gentoo-home/work/codes/moose/framework/src/utils/RankTwoTensor.x86_64-pc-linux-gnu.opt.lo] Error 1
```

I also updated and recompiled PETSc to version 3.16.5, OMPI 4.1.2 and GCC 11.2.1
Thanks for inputs.

--
Valmor

**Comment by GiudGiud** [comment, marked as answer]
Hi Valmor

looks like you just need to update libmesh. Please run ‘mamba update moose-libmesh’ and see if it fixes it. If it doesn’t then please report the output of ‘mamba list’

Guillaume

**Reply by dealmeidavf** [reply]
Hello Guillaume,

I do not have `mamba` installed on my `linux` system. Is that necessary?

libmesh has been fully recompiled with:` ./scripts/update_and_rebuild_libmesh.sh` issued in the `path-to/moose/` directory. Before that, the following `git` commands were used:

 • git clean -xfd
 • git submodule deinit -f libmesh
 • git submodule update --init --recursive libmesh

Thanks,

--
Valmor

**Reply by GiudGiud** [reply]
No it's not, using `update_and_rebuild_libmesh.sh` script is valid too.
But you dont need to initialize or update the libmesh submodule, the script does it for you. I'd deinit it then try the script again

**Reply by dealmeidavf** [reply]
I did:

 + git clean -xfd
 + git submodule deinit -f libmesh
 + ./scripts/update_and_rebuild_libmesh.sh
 + cd test
 + make -j 4

Still the same error; unable to compile tests or any app. I am not using tensors in my apps but still make will not work. Apparently if the test directory does not compile, apps will not compile either?
Thanks for inputs.

--
Valmor

**Reply by GiudGiud** [reply]
The error message seems to come from eigen, but not one we install, one that is already present on your system. 
Did you modify your LIBRARY_PATH environment variable?

**Reply by dealmeidavf** [reply]
No change to LIBRARY_PATH. I recompiled PETSc; is eigen from PETSc? I noticed I was not able to compile with `fblaslapack` and did not download it because I thought the Fortran version of blas and lapack are not used with c++ code; could this be the problem? That is, `fblaslapack` need to be downloaded with PETSc and compiled. The only reason to have PETSc on my system is to use it with MOOSE.

**Reply by GiudGiud** [reply]
oh the petsc is too new. 3.16 is not supported yet.

Could you please use the `update_abd_rebuild_petsc` script instead of installing it yourself?

**Reply by loganharbour** [reply]
PETSc 3.16.5 will be supported by the end of the week (likely sooner) in #20580.

**Reply by dealmeidavf** [reply]
I did recompile PETSc with `fblaslapack` (using the gitlab repo this turns out to be version 3.16.5) and recompiled the whole of MOOSE (libmesh included). The test build let to:

```bash
checking /usr/local/petsc_ompi//include/petscversion.h usability... no
checking /usr/local/petsc_ompi//include/petscversion.h presence... no
checking for /usr/local/petsc_ompi//include/petscversion.h... no
checking /usr/local/petsc_ompi/include/petscversion.h usability... no
checking /usr/local/petsc_ompi/include/petscversion.h presence... no
checking for /usr/local/petsc_ompi/include/petscversion.h... no
configure: error: *** PETSc was not found, but --enable-petsc-required was specified.
```

PETSc is definitely on my system, is this a version issue then? It would be nice to have a warning about the unsupported version if this is the case. Is there a file in the MOOSE directory that shows which version of PETSc is supported?
Thanks,

--
Valmor

**Reply by GiudGiud** [reply]
It is currently not supported. Please downgrade petsc if you want to work with MOOSE today.
To do that, use the update_and_rebuild_petsc script.

**Reply by GiudGiud** [reply]
Petsc is updated (has been for 2 weeks roughly)
so you can use this petsc now btw

**Reply by dealmeidavf** [reply]
I did a full update of PETSc on my system and a full `git pull` on `MOOSE` and 
`./scripts/update_and_rebuild_libmesh.sh --enable-vtk-required` gives:

```bash
<<< Using external boost installation >>>
checking for boost::movelib::unique_ptr support... yes
note: Checking /lib and /include for MPI
note: Could not find /lib/libmpi(.a/.so/.dylib)
mpicxx Compiler Supports MPI 
checking /usr/local/petsc_ompi//include/petscversion.h usability... no
checking /usr/local/petsc_ompi//include/petscversion.h presence... no
checking for /usr/local/petsc_ompi//include/petscversion.h... no
checking /usr/local/petsc_ompi/include/petscversion.h usability... no
checking /usr/local/petsc_ompi/include/petscversion.h presence... no
checking for /usr/local/petsc_ompi/include/petscversion.h... no
configure: error: *** PETSc was not found, but --enable-petsc-required was specified.
Running make -j 8...
make: *** No targets specified and no makefile found.  Stop.
```
Is this an issue with PETSc 3.17?

```bash
->  cat /usr/local/petsc_ompi//include/petscversion.h
#if !defined(PETSCVERSION_H)
#define PETSCVERSION_H
#include <petscconf.h>

#define PETSC_VERSION_RELEASE    0
#define PETSC_VERSION_MAJOR      3
#define PETSC_VERSION_MINOR      17
#define PETSC_VERSION_SUBMINOR   0
#define PETSC_RELEASE_DATE       "Mar 30, 2022"
#define PETSC_VERSION_DATE       "unknown"

```

Thanks.

**Reply by GiudGiud** [reply]
how did you install Petsc?
What does `echo $PETSC_ARCH $PETSC_DIR` return?

**Reply by GiudGiud** [reply]
oh btw we support petsc 3.16.5 not 3.17. Please downgrade

**Reply by dealmeidavf** [reply]
I have always compiled `MOOSE` using my local install of `PETSc` from:

```bash
->  git remote -v
origin  https://gitlab.com/petsc/petsc.git (fetch)
origin  https://gitlab.com/petsc/petsc.git (push)
```

I am wondering why I no longer can do this. Every time I pull `PETSc` (not often), I get a new version (now it is 3.7.0). I will try an older tarball and see what happens.

```bash
->  echo $PETSC_ARCH $PETSC_DIR
/usr/local/petsc_ompi

->  ls /usr/local/petsc_ompi/
include/  lib/  share/
```

The `MOOSE` script `./scripts/update_and_rebuild_libmesh.sh --enable-vtk-required` does not necessarily break because of an unsupported version; at least it is not explicitly written on the screen output.

**Reply by GiudGiud** [reply]
ok that is supported. But not 3.17, we support 3.16.5

**Reply by dealmeidavf** [reply]
I am not following your comment. The current version of `PETSc` from  

> https://gitlab.com/petsc/petsc.git (main branch) 

is: 3.17.0 which I have compiled from source and installed on my Linux box.

 Is this supposed to work with the current master branch of MOOSE at

```bash
->  git remote -v
origin  https://github.com/idaholab/moose (fetch)
origin  https://github.com/idaholab/moose (push)
```
 
If so, why is the MOOSE config script stopping at:

> configure: error: *** PETSc was not found, but --enable-petsc-required was specified.

PETSc is installed on my system. Why isn't the `MOOSE` configure script able to find `PETSc`? Is this a version issue? Is the `MOOSE` script looking for an older version of `PETSc`?

Thanks for inputs.

**Reply by GiudGiud** [reply]
no 3.17 is not supposed to work. 

Once you are on the supported version of Petsc you can try again. If `PETSC_DIR` and `PETSC_ARCH` are set, then it should know where to find Petsc
